### PR TITLE
Feature/Implement all previous features with the new `Instruction` data class

### PR DESCRIPTION
**Pull request by akshire**
# Implement all Previous Features with the New `Instruction` Data Class

## TLDR
```kotlin
instructions: String -> instructions: List<Instruction>
```

### Main Changes
This PR introduces a major change, which is the migration of the `Recipe` instructions from:
```kotlin
    val instructions: String,
```
to:
```kotlin
    val instructions: List<Instruction>
```

### Changes in Multiple Files

As a result, this PR introduces multiple changes in different screens that used the instructions of a `Recipe`, such as:
* `AddInstructionStepScreen`
* `RecipeListInstructionsScreen`
* `RecipeOverview` (major changes)

as well as backend classes used to manage `Recipe`s, including:
* `CreateRecipeViewModel` (logic modified)
* `FirestoreRecipesRepository`
* `MealDBRecipesRepository`
* `RecipeBuilder` (logic modified)

### Changes in `RecipeOverviewScreen`
Now, the instructions are displayed as a list of steps that are clickable. Clicking on them will display the description of the selected step.

https://github.com/user-attachments/assets/a14ee567-59ec-4850-aef9-8cfc839ea6d5

```kotlin
@Composable
private fun InstructionView(currentRecipe: Recipe) {
  Column(
      modifier =
          Modifier.padding(
                  start = OVERVIEW_INSTRUCTION_START.dp,
                  end = OVERVIEW_INSTRUCTION_END.dp,
                  top = OVERVIEW_INSTRUCTION_TOP.dp,
                  bottom = OVERVIEW_INSTRUCTION_BOTTOM.dp)
              .testTag(INSTRUCTIONS_VIEW)) {
        for (index in currentRecipe.instructions.indices) {
          InstructionValue(currentRecipe.instructions[index], index)
        }
      }
}
```
The `InstructionValue` function creates a `Card` with an `isExpanded` field that determines if the card is expanded or not.

### Changes in `CreateRecipeViewModel`
The changes made in this class mainly involve refactoring existing functions and implementing new functions that will be useful for the future implementation of the create `Recipe` procedure:

#### New Methods
```kotlin
fun resetSelectedInstruction()
fun getSelectedInstruction(): Int?
fun getInstruction(index: Int): Instruction
```

#### Refactored Methods
```kotlin
// Example of method refactoring: area -> origin
fun updateRecipeArea(area: String) -> fun updateRecipeOrigin(origin: String)

// Example of refactoring and adaptation
fun updateRecipeInstructions(instructions: String) -> fun addRecipeInstruction(instructions: Instruction)

// Example of refactoring
fun addIngredient(ingredient: String, measurement: String) -> fun addIngredientAndMeasurement(ingredient: String, measurement: String)
```

### Changes in `RecipeBuilder`
These changes are similar to the `CreateRecipeViewModel`. The major change is the instructions format change:
```kotlin
private var instructions: String = "" ->private var instructions: MutableList<Instruction> = mutableListOf()
```

Multiple methods needed adaptation like for example the `getInstructions()` method.

New methods were also introduced to enable modification and deletion of an instruction.

### Changes in Test Files
To be able to execute some tests, I had to modify certain structures within the tests. All the tests were preserved, but some logic was rethought (`CreateRecipeViewModel`), and the `Recipe` structure was updated.

### How to Test This
To test all these new changes, I had to create a new database to have the correct format of `Recipe`s. I will elaborate on this below.

## Regenerate the Database (not in the PR)
To test my implementation, I had to regenerate the entire database to match the correct format of existing recipes. The code below helped me achieve this. I reused some code from PR #201 written by @YlanV and @akshire (me).

### `MainActivity`
The function was added to the `MainActivity` class and is the main actor in this regeneration. This function generates a button that serves as confirmation to regenerate the entire database.
```kotlin
@Composable
fun TestButton() {
    var recipes = mutableListOf<Recipe>()
    Button(
        onClick = {
            val firestoreRecipesRepository = FirestoreRecipesRepository(Firebase.firestore)
            val imageRepository = ImageRepositoryFirebase(Firebase.storage)
            firestoreRecipesRepository.getAllRecipe(onSuccess = {
                recipes = it.toMutableList()
                Log.d("Test", recipes.size.toString())
                recipes.forEach { recipe ->
                    imageRepository.getImageUrl(recipe.uid, "Main",
                        ImageDirectoryType.RECIPE, onSuccess = { uri ->
                        recipe.url = uri.toString()
                        Log.d("Test", recipe.url!!)
                        firestoreRecipesRepository.addRecipe(recipe, onSuccess = {
                            Log.d("Test", "success")}, onFailure = {Log.d("Test", "error")})
                    }, onFailure = {
                        Log.d("Test", "error")
                    })
                }
            })
        }) {
        Text("Test")
    }
}
```
### Function Added to `FirestoreRecipesRepository`
This function was added to the `FirestoreRecipesRepository` to enable fetching the entire database and starting operations on it.

#### Warning
This function is safe to use for a small number of `Recipe`s (max 600).

```kotlin
fun getAllRecipe(onSuccess: (List<Recipe>) -> Unit) {
        db.collection("recipes with images")
            .get()
            .addOnSuccessListener { doc ->
                val recipes = mutableListOf<Recipe>()
                for (document in doc) {
                    val recipe = documentToRecipeOld(document)
                    if (recipe != null) {
                        recipes.add(recipe)
                        Log.d("FirestoreRecipesRepository", "getAllRecipe: ${recipe.name}")
                    }
                }
                onSuccess(recipes)
            }
            .addOnFailureListener { Log.e("FirestoreRecipesRepository", "Error getting documents", it) }
    }
```

### Function Added to `FirestoreRecipesRepository`
This function fetches `Recipe`s in the old format and was required to retrieve `Recipe`s from the old database.

```kotlin
fun documentToRecipeOld(document: DocumentSnapshot): Recipe? {
        return try {
            // Only check the mandatory arguments
            val id = document.id
            val name = document.getString(FIRESTORE_RECIPE_NAME) ?: return null
            val category = document.getString(FIRESTORE_RECIPE_CATEGORY)
            val area = document.getString(FIRESTORE_RECIPE_AREA)
            val instructions = document.getString("instructions") ?: return null
            val pictureID = document.getString(FIRESTORE_RECIPE_PICTURE_ID) ?: return null
            val ingredientsData = document.get(FIRESTORE_RECIPE_INGREDIENTS) as List<*>? ?: return null
            val measurementsData = document.get(FIRESTORE_RECIPE_MEASUREMENTS) as List<*>? ?: return null
            val time = document.getString(FIRESTORE_RECIPE_TIME)
            val difficulty = document.getString(FIRESTORE_RECIPE_DIFFICULTY)
            val price = document.getString(FIRESTORE_RECIPE_PRICE)
            val url = document.getString(FIRESTORE_RECIPE_URL)
            val ingredients = ingredientsData.mapNotNull { it as? String }
            val measurements = measurementsData.mapNotNull { it as? String }
            Recipe(
                uid = id,
                name = name,
                category = category,
                origin = area,
                instructions = listOf(Instruction(instructions)),
                strMealThumbUrl = pictureID,
                ingredientsAndMeasurements = ingredients.zip(measurements),
                time = time,
                difficulty = difficulty,
                price = price,
                url = url)
        } catch (e: Exception) {
            Log.e(
                "FirestoreRecipesRepository",
                "Error converting document to Recipe: idMeal " + document.id,
                e)
            null
        }
    }
```

### New Database: `Recipes with New Instructions`

I have created a new database that follows the updated format for `Recipe`s.

## Why Was This Change Made?

This change was made to allow users to separate the instructions of a `Recipe` into different steps, enabling a more fluid experience when reading the "How to Cook This Recipe" section.

## Checklist
- [x] Tests added.
- [x] Documentation updated.
- [x] All CI checks passed.
- [x] Linked issue/feature (if applicable): #205

### CODE REVIEW

File: app/src/main/java/com/android/sample/ui/recipeOverview/RecipeOverview.kt [hunk 1]
```diff
@@ -524,12 +536,72 @@ private fun InstructionView(currentRecipe: Recipe) {
...
+            verticalAlignment = Alignment.CenterVertically) {
+              Image(
+                  painter = painterResource(id = instruction.icon.iconResId),
+                  contentDescription = "Icon",
+                  modifier = Modifier.size(ICON_SIZE.dp).testTag("InstructionIcon"))
+
+              Column(modifier = Modifier.testTag("InstructionInfo")) {
+                Text(
+                    modifier = Modifier.testTag("InstructionTitle"),
+                    color = Color.Black,
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/ui/recipeOverview/RecipeOverview.kt:576]: **Hardcoded Colors:**
   - Avoid hardcoding colors like `Color.Black`. Use `MaterialTheme.colorScheme` instead to support light/dark themes dynamically.
   
   **Replace:**
   ```kotlin
   color = MaterialTheme.colorScheme.onPrimary
   ```

File: app/src/main/java/com/android/sample/ui/utils/SampleClasses.kt [hunk 2]
```diff
@@ -21,11 +22,33 @@ val recipeOrigins = arrayOf("Italy", "India", "United Kingdom", null)
...
     arrayOf(
-        "1. Boil water\n2. Add pasta\n3. Cook for 10 minutes\n4. Drain water\n5. Add sauce",
-        "1. Boil water\n2. Add chicken\n3. Cook for 10 minutes\n4. Drain water\n5. Add sauce",
-        "1. Boil water\n2. Add fish\n3. Cook for 10 minutes\n4. Drain water\n5. Add chips",
-        "take it out of the bag and shove it in your mouth",
-    )
+        listOf(
+            Instruction(description = "1. Boil water", time = "30 min", iconType = "Cook"),
+            Instruction(description = "2. Add pasta", time = "30 min", iconType = "Fire"),
+            Instruction(description = "3. Cook for 10 minutes", time = "30 min", iconType = "Fire"),
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/ui/utils/SampleClasses.kt:28]: Nothing serious, but it made me smile that in the instruction it says to cook for 10 min and that the time tag is 30 min

File: app/src/androidTest/java/com/android/sample/screen/RecipeOverviewTest.kt [hunk 3]
```diff
@@ -250,4 +251,65 @@ class RecipeOverviewTest {
...
+    composeTestRule.onNodeWithTag(INSTRUCTIONS_VIEW, useUnmergedTree = true).assertIsDisplayed()
+
+    val displayedinstructionsLength = 1 // adapted to small phone
+    val allInstructionsLength = testRecipes[0].instructions.size
+    composeTestRule
+        .onAllNodesWithTag("InstructionTitle", useUnmergedTree = true)
+        .assertCountEquals(allInstructionsLength)
+        .apply {
+          for (i in 0 until displayedinstructionsLength) {
+            Log.d("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA", "${get(i).toString()}")
```
Comment (2Fick) [inline app/src/androidTest/java/com/android/sample/screen/RecipeOverviewTest.kt:277]: don't forget to delete your log if you dont need them anymore. I know that testing is not an easy task 😢

File: app/src/main/java/com/android/sample/model/recipe/CreateRecipeViewModel.kt [hunk 4]
```diff
@@ -228,7 +273,7 @@ class CreateRecipeViewModel(
    *
    * @return The area of the recipe.
    */
-  fun getRecipeArea(): String? {
+  fun getRecipeOrigin(): String? {
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/recipe/CreateRecipeViewModel.kt:276]: You made the function name more descriptive. well done !

File: app/src/main/java/com/android/sample/model/recipe/CreateRecipeViewModel.kt [hunk 5]
```diff
@@ -210,10 +245,20 @@ class CreateRecipeViewModel(
    *
    * @return The instructions for the recipe.
    */
-  fun getRecipeInstructions(): String {
+  fun getRecipeInstructions(): List<Instruction> {
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/model/recipe/CreateRecipeViewModel.kt:248]: If this doesnt take too much time, is it possible to refactor the function name ? it looks very similar to the `getRecipeInstruction(index: Int): Instruction` one. I'd advise you to deal with the other points I raised in this review before doing this one, which I don't think is the most important yet.

File: app/src/main/java/com/android/sample/ui/createRecipe/AddInstructionStepScreen.kt [hunk 6]
```diff
@@ -215,14 +245,16 @@ fun confirmAndAssignStep(
     createRecipeViewModel: CreateRecipeViewModel,
     onSuccess: () -> Unit
 ) {
```
Comment (2Fick) [inline app/src/main/java/com/android/sample/ui/createRecipe/AddInstructionStepScreen.kt:247]: lgtm

File: app/src/main/java/com/android/sample/model/recipe/CreateRecipeViewModel.kt [hunk 7]
```diff
@@ -101,9 +127,18 @@ class CreateRecipeViewModel(
    *
    * @param instructions The new instructions for the recipe.
    */
-  fun updateRecipeInstructions(instructions: String) {
-    require(instructions.isNotBlank()) { "Instructions must not be blank." }
-    recipeBuilder.setInstructions(instructions)
+  fun addRecipeInstruction(instructions: Instruction) {
```
[the remark this answers is not in the document]
Comment (akshire) [reply app/src/main/java/com/android/sample/model/recipe/CreateRecipeViewModel.kt:130]: Yeah thanks for noticing that, I will do it right now

File: app/src/main/java/com/android/sample/ui/utils/SampleClasses.kt (hunk 2 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/ui/utils/SampleClasses.kt:28]: Yeah, I guess the one that wrote these tests must have been really tired

File: app/src/main/java/com/android/sample/model/recipe/Instruction.kt [hunk 8]
```diff
@@ -12,9 +12,13 @@ import com.android.sample.resources.C.TestTag.Instruction.DEFAULT_ICON
  */
 data class Instruction(
     val description: String,
-    val time: String?,
-    val iconType: String?,
+    val time: String? = "",
+    val iconType: String? = DEFAULT_ICON,
 ) {
```
[the remark this answers is not in the document]
Comment (akshire) [reply app/src/main/java/com/android/sample/model/recipe/Instruction.kt:17]: Nice catch!

File: app/src/main/java/com/android/sample/model/recipe/RecipeBuilder.kt [hunk 9]
```diff
@@ -145,7 +145,7 @@ class RecipeBuilder {
   fun build(): Recipe {
     // Validation for essential fields
     require(name.isNotBlank()) { ERROR_STR_MEAL_BLANK }
-    require(instructions.isNotBlank()) { ERROR_STR_INSTR_BLANK }
+    require(instructions.isNotEmpty()) { ERROR_STR_INSTR_BLANK }
```
[the remark this answers is not in the document]
Comment (akshire) [reply app/src/main/java/com/android/sample/model/recipe/RecipeBuilder.kt:148]: Thanks, we will discuss it tomorrow, but I think that it comes from the `Instruction` class where in the `init` method, I check if the `description` parameter is not blank. I also think it will take 2 minutes to fix.

File: app/src/main/java/com/android/sample/model/recipe/CreateRecipeViewModel.kt (hunk 5 above)
Comment (akshire) [reply app/src/main/java/com/android/sample/model/recipe/CreateRecipeViewModel.kt:248]: done

**Comment by akshire** [comment]
Hey I did all the changes, I ll let you analyze the tests I modified

**Review by 2Fick — APPROVED** [review]
Everythings looks good to me now !! Good job. 

PS: Can you just resolve the merge conflicts ?
